src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -656,14 +651,22 @@ export class TerminalService implements ITerminalService {
 	private _onWillShutdown(e: WillShutdownEvent): void {
 		// Don't touch processes if the shutdown was a result of reload as they will be reattached
 		const shouldPersistTerminals = this._configHelper.config.enablePersistentSessions && e.reason === ShutdownReason.RELOAD;
+		const terminals = [...this._terminalGroupService.instances, ...this._backgroundedTerminalInstances];
+
+		// Dispose of all transient terminals
+		for (const instance of terminals.filter(i => !i.shouldPersist)) {
+			instance.dispose(TerminalExitReason.Shutdown);
+		}
+
+		// Detach all persistent terminals
 		if (shouldPersistTerminals) {
-			for (const instance of this._terminalGroupService.instances) {
+			for (const instance of terminals.filter(i => i.shouldPersist)) {
 				instance.detachProcessAndDispose(TerminalExitReason.Shutdown);
 			}
 			return;
 		}
 
-		// Force dispose of all pane terminal instances
+		// Force dispose of all pane terminal instances as they will be recreated on restart
 		for (const instance of this._terminalGroupService.instances) {
 			instance.dispose(TerminalExitReason.Shutdown);
 		}
```

Comment:
I think this could now double dispose instances, how about instead of making `terminals`, you create `transientTerminals` and `persistentTerminals`?

We could also handle the `!shouldPersistTerminals` case first to avoid creating the `*Terminals` arrays when not needed.

Reply:
we check if it's been disposed before re-disposing, but sounds good to avoid that

https://github.com/microsoft/vscode/blob/bc1a3fdfc8c4d73fe7cb6a70aa1581cd49a2f259/src/vs/workbench/contrib/terminal/browser/terminalInstance.ts#L1184-L1187